### PR TITLE
[2.1] Remove AuthZ caching from AuthorizeFilter

### PR DESCRIPTION
**Pull request by HaoK**
2.1 Port of https://github.com/aspnet/AspNetCore/issues/9779

Customer Impact: When mixing multiple authentication schemes and authorize attributes (on controller + methods), we end up caching incorrectly, so the wrong authorization policy is applied on subsequent requests. This usually results in valid requests being rejected, I don't think there's any way this would let people who shouldn't be allowed access in.

Fix description: is to remove all the caching of authorization policies in MVC, so authorization policies are computed every time

Risk: Should be low (just removing caching logic, so just slower)

### CODE REVIEW

**Comment by anurse** [comment]
@rynowak @mkArtakMSFT can we get a review from someone on MVC (since this changes a bunch of AuthorizeFilter stuff). It's mostly a port from 3.0.

File: eng/PatchConfig.props [hunk 1]
```diff
@@ -35,5 +35,10 @@ Later on, this will be checked using this condition:
       Microsoft.AspNetCore.SignalR.Redis;
     </PackagesInPatch>
   </PropertyGroup>
+  <PropertyGroup Condition=" '$(VersionPrefix)' == '2.1.12' ">
+    <PackagesInPatch>
+      Microsoft.AspNetCore.Mvc.Core;
```
Comment (anurse) [inline eng/PatchConfig.props]: @aspnet/build please double-check this when you can :).

**Comment by HaoK** [comment]
Hold off on the review for a bit, looks like there were some 2.2 to 2.1 differences that I need to resolve, ah backporting fun

**Comment by rynowak** [comment]
OK. I'm ready to look at this when you are :)

**Comment by dougbu** [comment]
Do not merge until / unless this goes back to shiproom and is approved.

File: src/Mvc/Mvc.Core/src/Authorization/AuthorizeFilter.cs [hunk 2]
```diff
@@ -140,14 +152,13 @@ private async Task<AuthorizationPolicy> GetEffectivePolicyAsync(AuthorizationFil
                     
                     if (context.Filters[i] is AuthorizeFilter authorizeFilter)
                     {
-                        builder = builder ?? new AuthorizationPolicyBuilder(effectivePolicy);
-                        builder.Combine(authorizeFilter.Policy);
```
Comment (rynowak) [inline src/Mvc/Mvc.Core/src/Authorization/AuthorizeFilter.cs]: This logic is a little bit different right? `.Policy` is different stuff than `ComputePolicyAsync`

File: src/Mvc/test/Mvc.IntegrationTests/AuthorizeFilterIntegrationTest.cs [hunk 3]
```diff
@@ -72,18 +140,18 @@ private static ApplicationModelProviderContext GetProviderContext(Type controlle
...
-            serviceCollection.AddMvc();
+            serviceCollection.AddMvc(o => o.AllowCombiningAuthorizeFilters = combineAuthorize);
             serviceCollection
                 .AddTransient<ILoggerFactory, LoggerFactory>()
                 .AddTransient<ILogger<DefaultAuthorizationService>, Logger<DefaultAuthorizationService>>()
                 .AddSingleton<ObjectPoolProvider, DefaultObjectPoolProvider>();
             return serviceCollection.BuildServiceProvider();
         }
-
+
```
Comment (rynowak) [inline src/Mvc/test/Mvc.IntegrationTests/AuthorizeFilterIntegrationTest.cs]: Functional test with the repro for the bug that's being fixed?

File: src/Mvc/Mvc.Core/src/Authorization/AuthorizeFilter.cs (hunk 2 above)
Comment (rynowak) [reply src/Mvc/Mvc.Core/src/Authorization/AuthorizeFilter.cs]: I guess this is the same as what's in 2.2/3.0. Is the difference between these something to be concerned about?

Comment (HaoK) [reply src/Mvc/Mvc.Core/src/Authorization/AuthorizeFilter.cs]: Yeah 2.1 was different from 2.2/3.0 to start, I am trying to make all of them the same with this PR

Comment (rynowak) [reply src/Mvc/Mvc.Core/src/Authorization/AuthorizeFilter.cs]: I get that this changes the code to be more consistent, but does this change the observable behavior?

Comment (HaoK) [reply src/Mvc/Mvc.Core/src/Authorization/AuthorizeFilter.cs]: So maybe we shouldn't port this change to 2.1, it looks like we never ported this 2.2 fix to 2.1 which is why the code bases are so different...  https://github.com/aspnet/Mvc/commit/a5083d525b76ed8699d2b94e91d7ef52106d65aa#diff-fbf649a01b0521186bc5d75f07c20d8f

File: eng/PatchConfig.props (hunk 1 above)
Comment (natemcmaster) [reply eng/PatchConfig.props]: PatchConfig.props looks good now.

**Comment by HaoK** [comment]
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=213310

**Comment by HaoK** [comment]
I'm seeing some failures on Linux for 5 MVC Modelbinding tests (this should be totally unrelated to my changes except for maybe adding a reference to Mvc.Core).  Rerun didn't work either

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact aspnetcore-build@microsoft.com.

I've triaged the above build. I've created/commented on the following issue(s)
https://github.com/aspnet/AspNetCore-Internal/issues/2626

**Comment by HaoK** [comment]
Ok looks like the linux model binding failures are unrelated

File: src/Mvc/test/Mvc.IntegrationTests/AuthorizeFilterIntegrationTest.cs (hunk 3 above)
Comment (HaoK) [reply src/Mvc/test/Mvc.IntegrationTests/AuthorizeFilterIntegrationTest.cs]: I will try to port the new functional tests in 3.0 that pranav added for this scenario to 2.1/2.2 https://github.com/aspnet/AspNetCore/blob/master/src/Mvc/test/Mvc.FunctionalTests/AuthMiddlewareAndFilterTestBase.cs

**Comment by HaoK** [comment]
Closing for now

**Comment by udlose** [comment]
@HaoK was this ever fixed in .NET Core 2.2?  I am using 2.2 and experiencing this exact problem.

**Pull request closed without merging**